File: Data Structures & Algorithms/reorder-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* r = head;
        vector<ListNode> st;
        while (r != nullptr) {
            ListNode x = *r;
            x.next = nullptr;
            st.push_back(x);
            r = r->next;
        }

        for (auto& x : st) {
            cout << x.val << " ";
        }
        cout << endl;
        
        int len = st.size();

        ListNode* it = head;
        auto ac = new ListNode();
        auto res = ac;
        while (st.size() > (len/2)) {
            ac->next = it;
            it = it->next;
            ac = ac->next;
            ListNode* n = new ListNode(st.back().val);
            if (len % 2 == 1 && st.size() - (len/2) == 1) {
                ac->next = nullptr;
            } else {
                ac->next = n;
                ac = ac->next;
            }
            st.pop_back();
        }
        head = res->next;
    }
};

File: Data Structures & Algorithms/add-two-numbers/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        a = l1
        b = l2
        res = ListNode()
        x = res
        c = 0
        while a or b:
            s = c
            if a:
                s += a.val
                a = a.next
            if b:
                s += b.val
                b = b.next
            if s > 9:
                c = 1
                s -= 10
            else:
                c = 0
            res.next = ListNode(s)
            res = res.next
        
        if c >= 1:
            res.next = ListNode(c)
        
        return x.next
